stdlib/public/Concurrency/TaskSleep.swift

```diff
@@ -0,0 +1,144 @@
+  /// in nanoseconds.
+  ///
+  /// This function does _not_ block the underlying thread.
+  public static func sleep(_ duration: UInt64) async {
+    let currentTask = Builtin.getCurrentAsyncTask()
+    let priority = getJobFlags(currentTask).priority ?? Task.currentPriority._downgradeUserInteractive
+
+    return await Builtin.withUnsafeContinuation { (continuation: Builtin.RawUnsafeContinuation) -> Void in
+      let job = _taskCreateNullaryContinuationJob(priority: Int(priority.rawValue), continuation: continuation)
+      _enqueueJobGlobalWithDelay(duration, job)
+    }
+  }
+
+  /// The type of continuation used in the implementation of
+  /// sleep(nanoseconds:).
+  private typealias SleepContinuation = UnsafeContinuation<(), Error>
+
+  /// Called when the sleep(nanoseconds:) operation woke up without being
+  /// cancelled.
+  private static func onSleepWake(
+      _ wordPtr: UnsafeMutablePointer<Builtin.Word>,
+      _ continuation: UnsafeContinuation<(), Error>
+  ) {
+    // Indicate that we've finished by putting a "1" into the flag word.
+    let (_, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+        wordPtr._rawValue,
+        UInt(0)._builtinWordValue,
+        UInt(1)._builtinWordValue)
+
+    if Bool(_builtinBooleanLiteral: won) {
+      // The sleep finished, invoke the continuation.
+      continuation.resume()
+    } else {
+      // The task was cancelled first, which means the continuation was
+      // called by the cancellation handler. We need to deallocate up the flag
+      // word, because it was left over for this task to complete.
+      wordPtr.deallocate()
+    }
+  }
+
+  /// Called when the sleep(nanoseconds:) operation has been cancelled before
+  /// the sleep completed.
+  private static func onSleepCancel(
+      _ wordPtr: UnsafeMutablePointer<Builtin.Word>,
+      _ continuation: UnsafeContinuation<(), Error>
+  ) {
+    // Indicate that we've finished by putting a "2" into the flag word.
+    let (_, won) = Builtin.cmpxchg_seqcst_seqcst_Word(
+        wordPtr._rawValue,
+        UInt(0)._builtinWordValue,
+        UInt(2)._builtinWordValue)
+
+    if Bool(_builtinBooleanLiteral: won) {
+      // We recorded the task cancellation before the sleep finished, so
+      // invoke the continuation with a the cancellation error.
+      continuation.resume(throwing: _Concurrency.CancellationError())
+    }
+  }
+
+  /// Suspends the current task for _at least_ the given duration
+  /// in nanoseconds, unless the task is cancelled. If the task is cancelled,
+  /// throws \c CancellationError.
+  ///
+  /// This function does _not_ block the underlying thread.
+  public static func sleep(nanoseconds duration: UInt64) async throws {
+    // If the task was already cancelled, go ahead and throw now.
+    try checkCancellation()
+
+    // Allocate storage for the flag word and continuation.
+    let wordPtr = UnsafeMutablePointer<Builtin.Word>.allocate(capacity: 2)
```

Comment:
this word does not escape the context of this function right? it might be interesting to actually use a stack allocation instead of heap allocating it.

else wise if it does escape then this probably should be ref counted right?

Reply:
It does escape the context of this function: when the task is cancelled, the timed "wake" task has yet to be scheduled, and needs to check whether it finished first or not. In the case where it gets called but didn't finish first, it handles the deallocation.

No need for reference counting; the deallocation points are well-defined.